actors/builtin/payment_channel/payment_channel_actor.go

```diff
@@ -0,0 +1,247 @@
+				rt.Abort(8, "merged lane in voucher has outdated nonce, cannot redeem")
+			}
+
+			redeemedFromOthers = big.Add(redeemedFromOthers, otherls.Redeemed)
+			otherls.Nonce = merge.Nonce
+		}
+
+		// 2. To prevent double counting, remove already redeemed amounts (from
+		// voucher or other lanes) from the voucher amount
+		ls.Nonce = sv.Nonce
+		balanceDelta := big.Sub(sv.Amount, big.Add(redeemedFromOthers, ls.Redeemed))
+		// 3. set new redeemed value for merged-into lane
+		ls.Redeemed = sv.Amount
+
+		newSendBalance := big.Add(st.ToSend, balanceDelta)
+
+		// 4. check operation validity
+		if newSendBalance.LessThan(big.Zero()) {
+			rt.Abort(9, "voucher would leave channel balance negative")
+		}
+		if newSendBalance.GreaterThan(rt.CurrentBalance()) {
+			rt.Abort(10, "not enough funds in channel to cover voucher")
+		}
+
+		// 5. add new redemption ToSend
+		st.ToSend = newSendBalance
+
+		// update channel closingAt and MinCloseAt if delayed by voucher
+		if sv.MinCloseHeight != 0 {
+			if st.ClosingAt != 0 && st.ClosingAt < sv.MinCloseHeight {
+				st.ClosingAt = sv.MinCloseHeight
+			}
+			if st.MinCloseHeight < sv.MinCloseHeight {
+				st.MinCloseHeight = sv.MinCloseHeight
+			}
+		}
+		return nil
+	})
+	return &vmr.EmptyReturn{}
+}
+
+func (pca *PaymentChannelActor) Close(rt vmr.Runtime) *vmr.EmptyReturn {
+
+	var st PaymentChannelActorState
+	rt.State().Transaction(&st, func() interface{} {
+
+		rt.ValidateImmediateCallerIs(st.From, st.To)
+
+		if st.ClosingAt != 0 {
+			rt.Abort(1, "channel already closing")
+		}
+
+		st.ClosingAt = rt.CurrEpoch() + indices.PaymentChannel_PaymentChannelClosingDelay()
+		if st.ClosingAt < st.MinCloseHeight {
+			st.ClosingAt = st.MinCloseHeight
+		}
+
+		return nil
+	})
+	return &vmr.EmptyReturn{}
+}
+
+func (pca *PaymentChannelActor) Collect(rt vmr.Runtime) *vmr.EmptyReturn {
+
+	var st PaymentChannelActorState
+	rt.State().Transaction(&st, func() interface{} {
+
+		if st.ClosingAt == 0 {
+			rt.Abort(1, "payment channel not closing or closed")
+		}
+
+		if rt.CurrEpoch() < st.ClosingAt {
+			rt.Abort(2, "payment channel not closed yet")
+		}
+
+		// send remaining balance to "From"
+
+		_, code := rt.Send(
+			st.From,
+			builtin.MethodSend,
+			nil,
+			abi.NewTokenAmount(big.Sub(rt.CurrentBalance(), st.ToSend).Int64()),
+		)
+		vmr.RequireSuccess(rt, code, "Failed to send balance to `From`")
+
+		// send ToSend to "To"
+
+		_, code2 := rt.Send(
+			st.From,
+			builtin.MethodSend,
+			nil,
+			abi.NewTokenAmount(st.ToSend.Int64()),
+		)
+		vmr.RequireSuccess(rt, code2, "Failed to send funds to `To`")
+
+		st.ToSend = big.Zero()
+		return nil
+	})
+	return &vmr.EmptyReturn{}
+}
+
+func (pca *PaymentChannelActor) GetOwner(rt vmr.Runtime) addr.Address {
+	var st PaymentChannelActorState
+	rt.State().Readonly(&st)
+
+	return st.From
+}
+
+func (pca *PaymentChannelActor) GetToSend(rt vmr.Runtime) abi.TokenAmount {
+	var st PaymentChannelActorState
+	rt.State().Readonly(&st)
+
+	return st.ToSend
+}
```

Comment:
@anorth these aren't actually used in our code, but thought it useful to leave them as in Lotus. LMK if you want me to remove.

Reply:
I would prefer to remove them, but could be added to the state object instead.